### ISSUE TITLE
longer latency after post-training quantization

### ISSUE THREAD

**Issue by mengwanguc** [tensorflow/tensorflow#23759]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
I modified the example script a bit to print inference latency
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Google Colaboratory with CPU
- **TensorFlow installed from (source or binary)**:
! pip install -U tf-nightly==1.13.0.dev20181027
- **TensorFlow version (use command below)**:
tf-nightly-1.13.0.dev20181027
- **Python version**:
2.7
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

### Describe the problem
The post-training quantization documentation claims that it provides 3x lower latency: https://www.tensorflow.org/lite/performance/post_training_quantization

But when I try with the example script: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tutorials/post_training_quant.ipynb
I found that quantized model is apparently slower than original one.
(You can see the logs below. Quantized model provides longer latency).

Why is the experiment result different from the documentation claim?
Should I try some other example/environment?

### Source code / logs
The original one with floating-point:
Accuracy after 500 images: 0.970000 average time: 0.002977
Accuracy after 1000 images: 0.960000 average time: 0.003007
Accuracy after 1500 images: 0.952667 average time: 0.003022
Accuracy after 2000 images: 0.951500 average time: 0.003026
Accuracy after 2500 images: 0.949600 average time: 0.003026
Accuracy after 3000 images: 0.954000 average time: 0.003024
Accuracy after 3500 images: 0.956857 average time: 0.003029
Accuracy after 4000 images: 0.954000 average time: 0.003031
Accuracy after 4500 images: 0.954222 average time: 0.003032
Accuracy after 5000 images: 0.954200 average time: 0.003038
Accuracy after 5500 images: 0.957091 average time: 0.003043
Accuracy after 6000 images: 0.958500 average time: 0.003047
Accuracy after 6500 images: 0.959692 average time: 0.003050
Accuracy after 7000 images: 0.960714 average time: 0.003050
Accuracy after 7500 images: 0.962400 average time: 0.003051
Accuracy after 8000 images: 0.964375 average time: 0.003052
Accuracy after 8500 images: 0.965059 average time: 0.003051
Accuracy after 9000 images: 0.966889 average time: 0.003052
Accuracy after 9500 images: 0.968000 average time: 0.003052
Accuracy after 10000 images: 0.966700 average time: 0.003054
total time: 30.540090
0.9667

With post-training quantization:
Accuracy after 500 images: 0.970000 average time: 0.004113
Accuracy after 1000 images: 0.959000 average time: 0.004096
Accuracy after 1500 images: 0.950667 average time: 0.004093
Accuracy after 2000 images: 0.950000 average time: 0.004084
Accuracy after 2500 images: 0.948000 average time: 0.004077
Accuracy after 3000 images: 0.952667 average time: 0.004081
Accuracy after 3500 images: 0.955714 average time: 0.004103
Accuracy after 4000 images: 0.953000 average time: 0.004125
Accuracy after 4500 images: 0.953111 average time: 0.004133
Accuracy after 5000 images: 0.953200 average time: 0.004130
Accuracy after 5500 images: 0.956182 average time: 0.004126
Accuracy after 6000 images: 0.957667 average time: 0.004132
Accuracy after 6500 images: 0.958923 average time: 0.004137
Accuracy after 7000 images: 0.960000 average time: 0.004142
Accuracy after 7500 images: 0.961600 average time: 0.004146
Accuracy after 8000 images: 0.963625 average time: 0.004149
Accuracy after 8500 images: 0.964353 average time: 0.004148
Accuracy after 9000 images: 0.966222 average time: 0.004145
Accuracy after 9500 images: 0.967368 average time: 0.004140
Accuracy after 10000 images: 0.966100 average time: 0.004136
total time: 41.360908
0.9661

We can see after post-training quantization, the latency is apparently longer......
But in the documentation, it claims to reduce latency....

**Comment by damhurmuller** [comment]
Same problem here. I trained mobilenet on keras on my own dataset and after convert to tflite and use on the application, the FPS was around 250ms and with the post-training quantization was about 450ms.

**Keras:** 2.2.4
**tf-nightly-gpu**: '1.13.0-dev20181110'

**Comment by damhurmuller** [comment]
@mengwanguc did you find any solution?

**Comment by minizon** [comment]
Same problem.

**Comment by suharshs** [comment]
Hi all, thanks for the issue!
We may need to optimize some kernels that your models are using. We experienced up to 3x on models we optimized, but there are many ops that may still need extra attention :)

We are looking into this. To help us reproduce, could you provide the following:
(1) Which model are you running?
(2) What device/architecture are you running on?

Also, if possible could you provide your TFLite float and TFLite quantized model for reproducing? Feel free to direct email me.

**Comment by mengwanguc** [comment]
Hi @suharshs 

(1). I'm using the MNIST model
(2). I'm running on Google Colab with CPU.
       It's Intel(R) Xeon(R) CPU @ 2.30GHz with ~12.7G RAM.

Here the generated TFLite models :
TFLite float: https://drive.google.com/file/d/146fFE4jEXaTLfgXWK3ByQ0XOxIYqa8I1/view?usp=sharing
TFLite quantized: https://drive.google.com/file/d/1dcl9hWzcKNSvoOQBqgTOy3QLsvJLz010/view?usp=sharing

Thanks,
Meng

**Comment by mengwanguc** [comment]
@suharshs 
Actually this is the Google Colab notework I used: https://colab.research.google.com/drive/1ki6VI522w8oUCdoTu4azPZ2_dSZMPHtB

**Comment by jazzystring1** [comment]
Same problem here

**Comment by mengwanguc** [comment]
OK I guess the 10x speed-up is caused by num_threads.

In my previous experiment, I followed the benchmark document, and set the num_threads to 4. The speed-up was 10x.

I changed num_threads to 1, and the results are similar to yours:
float model: avg=5347.93us
quant model: avg=2734.78us

Looks like quant model does better for multi-threading?
I have some guess: is that because quant model requires less memory access, thus when running with multiple threads, the queueing for shared memory (model weights) takes less time?
